lib/ansible/modules/windows/win_file.ps1

```diff
@@ -20,51 +20,601 @@
+            }
+            else
+            {
+                linkFlags = SYMBOLIC_LINK_FLAG_FILE;
+            }
+
+            bool success = CreateSymbolicLink(linkPath, target, linkFlags);
+            if (!success)
+                throw new Exception(String.Format("Failed to create Symbolic Link {0} to Target {1}: {2}", linkPath, target, new Win32Exception(Marshal.GetLastWin32Error()).Message));
+        }
+
+        private static void CreateLinkHard(string linkPath, String target)
+        {
+            bool success = CreateHardLink(linkPath, target, IntPtr.Zero);
+            if (!success)
+                throw new Exception(String.Format("Failed to create Hard Link {0} to Target {1}: {2}", linkPath, target, new Win32Exception(Marshal.GetLastWin32Error()).Message));
+        }
+
+        private static void CreateJunctionPoint(string linkPath, string target)
+        {
+            Directory.CreateDirectory(linkPath);
+            SafeFileHandle handle = CreateFile(linkPath, FileAccess.Write, FileShare.Read | FileShare.Write | FileShare.None, IntPtr.Zero, FileMode.Open, FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OPEN_REPARSE_POINT, IntPtr.Zero);
+            if (handle.IsInvalid)
+                throw new Exception(String.Format("Unable to open Reparse Point at {0}: {1}", linkPath, new Win32Exception(Marshal.GetLastWin32Error()).Message));
+
+
+            string substituteName = MAXIMUM_PATH_FIX_PREFIX + Path.GetFullPath(target);
+            string printName = target;
+
+            short substituteNameOffset = 0;
+            short substituteNameLength = (short)(substituteName.Length * SIZE_OF_WCHAR);
+
+            short printNameOffset = (short)(substituteNameLength + 2);
+            short printNameLength = (short)(printName.Length * SIZE_OF_WCHAR);
+
+            short reparseDataLength = (short)(substituteNameLength + printNameLength + 12);
+
+            byte[] unicodeBytes = Encoding.Unicode.GetBytes(substituteName + "\0" + printName);
+            char[] pathBuffer = Encoding.Unicode.GetChars(unicodeBytes);
+
+            REPARSE_DATA_BUFFER reparseDataBuffer = new REPARSE_DATA_BUFFER();
+            reparseDataBuffer.ReparseTag = IO_REPARSE_TAG_MOUNT_POINT;
+            reparseDataBuffer.ReparseDataLength = reparseDataLength;
+            reparseDataBuffer.SubstituteNameOffset = substituteNameOffset;
+            reparseDataBuffer.SubstituteNameLength = substituteNameLength;
+            reparseDataBuffer.PrintNameOffset = printNameOffset;
+            reparseDataBuffer.PrintNameLength = printNameLength;
+            reparseDataBuffer.PathBuffer = new char[MAXIMUM_REPARSE_DATA_BUFFER_SIZE];
+
+            Array.Copy(pathBuffer, reparseDataBuffer.PathBuffer, pathBuffer.Length);
+            int inBufferSize = Marshal.SizeOf(reparseDataBuffer);
+            IntPtr inBuffer = Marshal.AllocHGlobal(inBufferSize);
+
+            try
+            {
+                Marshal.StructureToPtr(reparseDataBuffer, inBuffer, false);
```

Comment:
At first glance, all the usages of `REPARSE_DATA_BUFFER` appear to be blittable- since we're not using DeviceIoControl for anything else, we could just change the arg to `ref REPARSE_DATA_BUFFER` and ditch all the manual allocation/free stuff entirely with some tweaks to the struct.